### PR TITLE
Remove index multiplying to the sizeof(WCHAR) after EvtFormatMessage in win32evtlog.i file.

### PR DESCRIPTION
**Pull request by ArthurKamalov**
This small PR removes index multiplying to the `sizeof(WCHAR)`. According to the [docs](https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtformatmessage), the `EvtFormatMessage` function does not treat the buffer size as amount bytes but as the amount of characters.

### CODE REVIEW

**Comment by ArthurKamalov** [comment]
@mhammond I know that you are in the process of dropping python 2 support and there are a lot of things to do so, I really don't want to rush you, but our product highly depends on Windows event log API and it suffers from constant crashes which should be caused by wrong buffer indexing. It would be super awesome if you could have a look at it and give your feedback. Now we have to build pywin32 from this fork so, if there are no problems, we would be very happy to see this fix in the next release :)


Thanks in advance.


PS. All CI jobs fail but it seems like this is not connected with my changes. I think there is something with `PythonServiceMessages.h` file. Looks like it has been removed but there are still references to it.

File: win32/src/win32evtlog.i [hunk 1]
```diff
@@ -562,8 +562,7 @@ static PyObject *FormatMessageInternal(EVT_HANDLE metadata, EVT_HANDLE event, DW
 		allocated_size += 1; // +1 to double terminate the keyword list
 	}
 
-	allocated_size *= sizeof(WCHAR);
-	buf = (WCHAR *)malloc(allocated_size);
+	buf = (WCHAR *)malloc(allocated_size * sizeof(WCHAR));
```
Comment (mhammond) [inline win32/src/win32evtlog.i:565]: why this change? `allocated_size` is no longer the allocated size.

File: win32/src/win32evtlog.i [hunk 2]
```diff
@@ -579,7 +578,7 @@ static PyObject *FormatMessageInternal(EVT_HANDLE metadata, EVT_HANDLE event, DW
 	}
 
 	if (flags == EvtFormatMessageKeyword) {
-		buf[returned_size * sizeof(WCHAR)] = L'\0';
+		buf[returned_size - 1] = L'\0';
```
Comment (mhammond) [inline win32/src/win32evtlog.i:581]: This doesn't make sense to me - this is treating `returned_size` as the number of bytes and not chars, which is the opposite of what we need to do. IOW, this will be placing a \0 in the middle of the allocated buffer not at the end.

**Review by mhammond — COMMENTED** [review]
I don't understand this - what problems do you see using it without these changes?

File: win32/src/win32evtlog.i (hunk 2 above)
Comment (ArthurKamalov) [reply win32/src/win32evtlog.i:581]: According to docs https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtformatmessage

> _BufferUsed_
The size, in **characters** of the caller-allocated buffer that the function used or the required buffer size if the function fails with ERROR_INSUFFICIENT_BUFFER.


so the `returned_size ` variable, which is given as a `BufferUsed` parameter, should be an amount of characters, shouldn't it?

Without those changes, our app had been constantly crashing, and that's(as I assume) why:
* `EvtFormatMessage` sets some value for the `returned_size`, for example `10` characters.

* Then we allocate memory for 10 **bytes**, not characters.
_This is a previous code._
```
	allocated_size *= sizeof(WCHAR);
	buf = (WCHAR *)malloc(allocated_size);
```
So we have a buffer with a size - 10 bytes, but the size of the string is 10 wide characters(20 bytes)
* Then:
_This is a previous code._
```
buf[returned_size * sizeof(WCHAR)] = L'\0';
```
we try to set zero termination character by index 20 in the buffer with size 10 bytes. The result - segfault.

It also seems like we should `returned_size - 1`. You can see that in the official [docs](https://docs.microsoft.com/en-us/windows/win32/wes/formatting-event-messages) too.

Comment (ArthurKamalov) [reply win32/src/win32evtlog.i:581]: We should not have to do  `returned_size - 1`, because we increment `allocated_size`, but I am still sure that we should not multiply `returned_size` to `sizeof(WCHAR)`. 

According to docs https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtformatmessage

>BufferUsed
The size, in characters of the caller-allocated buffer that the function used or the required buffer size if the function fails with ERROR_INSUFFICIENT_BUFFER.

so the `returned_size` variable, which is given as a `BufferUsed` parameter, should be a number of characters, shouldn't it?

Considering that, we don't have to multiply indices because the type of the `buf` variable is `LPWSTR`, not just a raw bytes array. If we do multiply, we go out of bufffer's bounds. Sometimes it does not crash because `malloc` may allocate more bytes but  this is still undefined behaviour.

File: win32/src/win32evtlog.i (hunk 1 above)
Comment (ArthurKamalov) [reply win32/src/win32evtlog.i:565]: Yes, I understand my mistake now. After my changes, the `allocated_size` had a misleading name and there was no need for these changes in `allocated_size` and malloc at all. So I'll return back this part of the code as it was.

**Comment by ArthurKamalov** [comment]
@mhammond Thank you for your reply and thanks for pointing on my mistakes. I was fast enough to reply to you with some comments, but I deleted them because they were wrong.  Please ignore them.

I've returned back everything exept buffer indexing  multiplying. You can see why in the [previous comment.](https://github.com/mhammond/pywin32/pull/1577#discussion_r491041288)

File: win32/src/win32evtlog.i [hunk 3]
```diff
@@ -579,7 +579,7 @@ static PyObject *FormatMessageInternal(EVT_HANDLE metadata, EVT_HANDLE event, DW
 	}
 
 	if (flags == EvtFormatMessageKeyword) {
-		buf[returned_size * sizeof(WCHAR)] = L'\0';
+		buf[returned_size] = L'\0';
```
Comment (mhammond) [inline win32/src/win32evtlog.i:582]: This still seems wrong to me, but the MS docs aren't clear, and in some cases, simply wrong.

`returned_size` being characters means this is still going to be sticking a null in the middle of the string. Let's say `returned_size` comes back as 10 - this means we need a buffer for 10 wide chars, so 20 bytes. We'll correctly allocate the 20 bytes, but here we will stick the \0 10 *bytes* into that buffer. Up at line 561 we've ensured that the buffer is large enough to take this extra \0, and we've done this before multiplying by `sizeof(WCHAR)` - so this looks perfectly correct to me as it stands, and your change will stick a \0 in the middle of the buffer and may not leave a double-null terminated string.

(https://docs.microsoft.com/en-us/windows/win32/wes/formatting-event-messages is about the only example I can find trying to use this, but it's clearly wrong in that it does:

```
            if ((EvtFormatMessageKeyword == FormatId))
                pBuffer[dwBufferSize-1] = L'\0';
            else
                dwBufferSize = dwBufferUsed;
```
when `pBuffer` is null and `dwBufferSize` is zero!

The only other possibility I can see is that `returned_size` does not count the "normal" trailing \0 character - ie, if the returned string was going to be 'X', then `returned_size` might return 1 - meaning we need to allocate space for *2* chars (ie, for `X\0`) and *3* chars for `EvtFormatMessageKeyword` (so we end up with `X\0\0` in that case). Does you crash go away if line 560 changes to `allocated_size = returned_size + 1;`?

There aren't any samples which use `EvtFormatMessageKeyword` - can you maybe share one?

**Review by mhammond — CHANGES_REQUESTED** [review]

File: win32/src/win32evtlog.i (hunk 3 above)
Comment (ArthurKamalov) [reply win32/src/win32evtlog.i:582]: >returned_size being characters means this is still going to be sticking a null in the middle of the string. Let's say returned_size comes back as 10 - this means we need a buffer for 10 wide chars, so 20 bytes. We'll correctly allocate the 20 bytes, but here we will stick the \0 10 bytes into that buffer.

But what about that the type of the `buf` is `LPWSTR`? Does that mean that when we do `buf[returned_size]`, we access 10th two-byte long WCHAR **element**, not byte, and this element should be located in the 19th and 20th byte respectively?

Comment (mhammond) [reply win32/src/win32evtlog.i:582]: doh! Yes, you are correct.

**Review by mhammond — APPROVED** [review]

**Comment by ArthurKamalov** [comment]
Thanks a lot!
